draft-ietf-tls-tls13.md

```diff
@@ -2810,32 +2827,82 @@ configuration identifier for each connection, thus preventing
 replay. Implementations are responsible for ensuring uniqueness of the
 identifier in this case.
 
+The "ticket_age" extension sent by the client SHOULD be used by
+servers to limit the time over which the first flight might be
+replayed.  A server can store the time at which it sends a server
+configuration to a client, or encode the time in a ticket.  Then, each
+time it receives an early_data extension, it can check to see if the
+value used by the client matches its expectations.
+
+The "ticket_age" value provided by the client will be shorter than the
+actual time elapsed on the server by a single round trip time.  This
+difference is comprised of the delay in sending the NewSessionTicket
+message to the client, plus the time taken to send the ClientHello to
+the server.  For this reason, a server SHOULD measure the round trip
+time prior to sending the NewSessionTicket message and account for
+that in the value it saves.
+
+There are several potential sources of error that make an exact
+measurement of time difficult.  Variations in client and server clocks
+are likely to be minimal, outside of gross time corrections.  Network
+propagation delays are most likely causes of a mismatch in legitimate
+values for elapsed time.  Both the NewSessionTicket and ClientHello
+messages might be retransmitted and therefore delayed, which might be
+hidden by TCP.
+
+A small allowance for errors in clocks and variations in measurements
+is advisable.  However, any allowance also increases the opportunity
+for replay.  In this case, it is better to reject early data than to
+risk greater exposure to replay attacks.
+
+#### Ticket Age
+
+%%% Key Exchange Messages
+       struct {
+           uint64 ticket_age;
+       } TicketAge;
+
+When the client sends the "early_data" extension, it MUST also send
+a "ticket_age" extension in its EncryptedExtensions block. This value
+contains the time elapsed since the client learned about the server
+configuration that it is using, in milliseconds.  This value can
+be used by the server to limit the time over which early data can
+be replayed.
```

Comment:
Most likely Martin's text. I don't think I've touched any of the text in this area of the spec. (My guess is you just didn't notice that one of the comments here wasn't his. ;)

And yeah, 'tis a bikesheddy topic, but without any actual numbers here, my worry is that some implementation will happily accept a replay a year and a half after the fact.

To be more precise, I'm asking for about 4 values here: recommended & hard max sanity checks, for both acceptable error & total age. Changing the field to an unsigned 32-bit int, instead of 64-bit, would set an effectively unavoidable max at a bit under 50 days. (2^32-1 ms ~= 49.7 days)


On Wednesday, May 04, 2016 09:03:54 am ekr wrote:
> Me too, but this is your text. Recommendations wanted


On Wednesday, May 04, 2016 10:41:27 am Martin Thomson wrote:
> I think that it's my text.  And I sense a really big bike shed looming.
> 
> If I had to pick, I'd say 100ms, but I don't know how good RTT estimates are.